### PR TITLE
CAMEL-10178: Google PubSub Component

### PR DESCRIPTION
**Pull request by evmin**
Google PubSub component with producer and consumer endpoints, unit and integration tests.

### CODE REVIEW

File: components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+import org.apache.camel.spi.UriParam;
+import org.apache.camel.spi.UriPath;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * PubSub Endpoint Definition
+ */
+@UriEndpoint(scheme = "google-pubsub", title = "Google Pubsub",
+        syntax = "google-pubsub:projectId:destinationName?[options]", label = "messaging")
```
Comment (davsclaus) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: The syntax should be without [options]. Those are implied.

File: components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    private String projectId;
+
+    @UriPath(description = "Destination Name")
+    @Metadata(required = "true")
+    private String destinationName;
+
+    @UriParam(name = "loggerId", description = "Logger ID to use when a match to the parent route required")
+    private String loggerId;
+
+    @UriParam(name = "concurrentConsumers", description = "The number of parallel streams consuming from the subscription")
```
Comment (davsclaus) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: You need to set defaultValue in the `@UriParam` to 1 so its the same as the field value

File: components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    private Integer concurrentConsumers = 1;
+
+    @UriParam(name = "maxMessagesPerPoll", description = "The max number of messages to receive from the server in a single API call")
+    private Integer maxMessagesPerPoll = 1;
+
+    @UriParam(name = "connectionFactory", description = "ConnectionFactory to obtain connection to PubSub Service. If non provided the default one will be used")
+    private GooglePubsubConnectionFactory connectionFactory;
+
+    @UriParam(defaultValue = "AUTO", enums = "AUTO, NONE",
```
Comment (davsclaus) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: There should not be space in the `enums` attribute, eg `AUTO,NONE`

File: components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java (hunk 1 above)
Comment (evmin) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: Thank you. Done.

File: components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java (hunk 2 above)
Comment (evmin) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: Done for both params.

File: components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java (hunk 3 above)
Comment (evmin) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: Done.

**Comment by evmin** [comment]
@davsclaus Once again - thank you. Anything else?

**Comment by davsclaus** [comment]
Seems good. I think @oscerd and others have helped you review recently. So I will let them merge the PR.

File: components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+import org.apache.camel.Producer;
+import org.apache.camel.impl.DefaultEndpoint;
+import org.apache.camel.spi.Metadata;
+import org.apache.camel.spi.UriEndpoint;
+import org.apache.camel.spi.UriParam;
+import org.apache.camel.spi.UriPath;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
```
Comment (davsclaus) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: Ah yeah javadoc on endpoint become the component description AFAIR in the generated docs. So please updated this.  eg its used for the description summary at https://github.com/apache/camel/tree/master/components#components

Comment (evmin) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: Gladly. What would the component to demonstrate a good example of such doco?

Comment (davsclaus) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: For example
https://github.com/apache/camel/blob/master/components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java#L38

Comment (evmin) [inline components/camel-google-pubsub/src/main/java/org/apache/camel/component/google/pubsub/GooglePubsubEndpoint.java]: Thank you. Done.

**Comment by oscerd** [comment]
Thanks the PR has been merged. I had to fix a little thing here:

https://github.com/apache/camel/commit/e3bc3ced52eeb7e4dbf620b9c77c2084c16deb75

Can you close the PR?

Many thanks!

**Pull request closed without merging**